homeassistant/components/websocket_api/automation.py

```diff
@@ -0,0 +1,151 @@
+
+    @classmethod
+    def create(cls, component: str, target_description: dict[str, Any]) -> Self:
+        """Build automation component lookup data from target description."""
+        filters: list[_EntityFilter] = []
+
+        entity_filters_config = target_description.get("entity", [])
+        for entity_filter_config in entity_filters_config:
+            entity_filter = _EntityFilter(
+                domains=set(entity_filter_config.get("domain", [])),
+                device_classes=set(entity_filter_config.get("device_class", [])),
+                supported_features=set(
+                    entity_filter_config.get("supported_features", [])
+                ),
+            )
+            filters.append(entity_filter)
+
+        return cls(component=component, filters=filters)
+
+    def matches(self, hass: HomeAssistant, entity_id: str, domain: str) -> bool:
+        """Return if entity matches ANY of the filters."""
+        if not self.filters:
+            return True
+        return any(f.matches(hass, entity_id, domain) for f in self.filters)
+
+
+def _async_get_automation_components_for_target(
+    hass: HomeAssistant,
+    target_selection: ConfigType,
+    expand_group: bool,
+    component_descriptions: dict[str, dict[str, Any] | None],
+) -> set[str]:
+    """Get automation components (triggers/conditions/services) for a target.
+
+    Returns all components that can be used on any entity that are currently part of a target.
+    """
+    extracted = target_helpers.async_extract_referenced_entity_ids(
+        hass,
+        target_helpers.TargetSelectorData(target_selection),
+        expand_group=expand_group,
+    )
+    _LOGGER.debug("Extracted entities for lookup: %s", extracted)
+
+    # Build lookup structure: domain -> list of trigger/condition/service lookup data
+    domain_components: dict[str, list[_AutomationComponentLookupData]] = {}
+    component_count = 0
+    for component, description in component_descriptions.items():
+        if description is None or CONF_TARGET not in description:
+            _LOGGER.debug("Skipping component %s without target description", component)
+            continue
+        domain = component.split(".")[0]
+        domain_components.setdefault(domain, []).append(
+            _AutomationComponentLookupData.create(component, description[CONF_TARGET])
+        )
+        component_count += 1
+
+    _LOGGER.debug("Automation components per domain: %s", domain_components)
+
+    entity_infos = entity_sources(hass)
+    matched_components: set[str] = set()
+    for entity_id in extracted.referenced | extracted.indirectly_referenced:
+        if component_count == len(matched_components):
+            # All automation components matched already, so we don't need to iterate further
+            break
+
+        entity_info = entity_infos.get(entity_id)
+        if entity_info is None:
+            _LOGGER.warning("No entity source found for %s", entity_id)
```

Comment:
Why the warning? When does this happen?

Reply:
OK, and why do we log a warning? Does it mean we'll spam warnings when editing a service call or trigger in an automation which is referencing a deleted entity for example? I don't think that's very helpful, although I think we might want to tell frontend about it (but not in this PR).